File: frontend/apps/ios/Pantopus/Features/Feed/Pulse/PulsePostCard.swift
//
//  PulsePostCard.swift
//  Pantopus
//
//  One row in the Pulse vertical feed. Header (avatar + meta + intent
//  chip), optional title (Event only), body clamped to 3 lines (2 with
//  a title), intent-shaped reaction strip, optional Event attendee row
//  with RSVP chip. Pure-render view — all state lives upstream.
//

import SwiftUI

/// VM-prepared content for a single Pulse card.
public struct PulsePostCardContent: Sendable, Hashable, Identifiable {
    public let id: String
    public let authorName: String
    public let authorInitials: String
    public let authorVerified: Bool
    public let avatarTint: FeedAvatarTint
    public let meta: String
    public let intent: PulseIntent
    public let title: String?
    public let body: String
    public let reactions: [PulseReaction]
    public let attendees: PulseAttendeeStrip?
    public let userHasReacted: Bool
    public let media: [PostMediaItem]
    /// Number of comments — shown beside the Reply affordance.
    public let commentCount: Int

    /// Still-image URLs — kept for call sites (and tests) that only care
    /// about what the card displays, not the attachment kinds.
    public var mediaURLs: [URL] {
        media.map(\.url)
    }

    public init(
        id: String,
        authorName: String,
        authorInitials: String,
        authorVerified: Bool,
        avatarTint: FeedAvatarTint = .sky,
        meta: String,
        intent: PulseIntent,
        title: String?,
        body: String,
        reactions: [PulseReaction],
        attendees: PulseAttendeeStrip?,
        userHasReacted: Bool,
        media: [PostMediaItem] = [],
        commentCount: Int = 0
    ) {
        self.id = id
        self.authorName = authorName
        self.authorInitials = authorInitials
        self.authorVerified = authorVerified
        self.avatarTint = avatarTint
        self.meta = meta
        self.intent = intent
        self.title = title
        self.body = body
        self.reactions = reactions
        self.attendees = attendees
        self.userHasReacted = userHasReacted
        self.media = media
        self.commentCount = commentCount
    }
}

/// Event card attendee strip — stacked avatars + going count + RSVP CTA.
public struct PulseAttendeeStrip: Sendable, Hashable {
    public let avatars: [String]
    public let goingCount: Int
    public let userIsGoing: Bool

    public init(avatars: [String], goingCount: Int, userIsGoing: Bool) {
        self.avatars = avatars
        self.goingCount = goingCount
        self.userIsGoing = userIsGoing
    }
}

/// Pulse post card — entirely render-only; tap dispatch is parent-controlled.
public struct PulsePostCard: View {
    private let content: PulsePostCardContent
    private let onTap: @MainActor () -> Void
    private let onPrimaryReaction: @MainActor () -> Void
    private let onRSVP: (@MainActor () -> Void)?

    public init(
        content: PulsePostCardContent,
        onTap: @escaping @MainActor () -> Void,
        onPrimaryReaction: @escaping @MainActor () -> Void,
        onRSVP: (@MainActor () -> Void)? = nil
    ) {
        self.content = content
        self.onTap = onTap
        self.onPrimaryReaction = onPrimaryReaction
        self.onRSVP = onRSVP
    }

    public var body: some View {
        Button(action: onTap) {
            VStack(alignment: .leading, spacing: Spacing.s2) {
                header
                if let title = content.title, !title.isEmpty {
                    Text(title)
                        .font(.system(size: 13.5, weight: .semibold))
                        .foregroundStyle(Theme.Color.appText)
                        .lineLimit(1)
                }
                if !content.body.isEmpty {
                    Text(content.body)
                        .font(.system(size: 12.5))
                        .foregroundStyle(Theme.Color.appTextStrong)
                        .lineLimit(content.title?.isEmpty == false ? 2 : 3)
                        .frame(maxWidth: .infinity, alignment: .leading)
                }
                if !content.media.isEmpty {
                    PostMediaGridView(
                        items: content.media,
                        style: .compact,
                        accessibilityID: "pulsePostMedia_\(content.id)"
                    )
                }
                if let attendees = content.attendees {
                    attendeeStrip(attendees)
                }
                reactionStrip
            }
            .padding(Spacing.s3)
            .background(Theme.Color.appSurface)
            .clipShape(RoundedRectangle(cornerRadius: Radii.xl, style: .continuous))
            .overlay(
                RoundedRectangle(cornerRadius: Radii.xl, style: .continuous)
                    .stroke(Theme.Color.appBorder, lineWidth: 1)
            )
        }
        .buttonStyle(.plain)
        .accessibilityElement(children: .combine)
        .accessibilityLabel(a11yLabel)
        .accessibilityIdentifier("pulsePostCard_\(content.id)")
    }

    private var header: some View {
        HStack(alignment: .center, spacing: 9) {
            FeedAvatar(
                initials: content.authorInitials,
                tint: content.avatarTint,
                verified: content.authorVerified,
                size: 32
            )
            VStack(alignment: .leading, spacing: 2) {
                Text(content.authorName)
                    .font(.system(size: 13, weight: .semibold))
                    .foregroundStyle(Theme.Color.appText)
                    .lineLimit(1)
                Text(content.meta)
                    .font(.system(size: 10.5))
                    .foregroundStyle(Theme.Color.appTextSecondary)
                    .lineLimit(1)
            }
            Spacer(minLength: Spacing.s2)
            PulseIntentChip(intent: content.intent)
        }
    }

    private func attendeeStrip(_ strip: PulseAttendeeStrip) -> some View {
        VStack(spacing: Spacing.s0) {
            Rectangle()
                .fill(Theme.Color.appBorderSubtle)
                .frame(height: 1)
                .padding(.top, Spacing.s2)
            HStack(alignment: .center, spacing: Spacing.s2) {
                HStack(spacing: -8) {
                    ForEach(Array(strip.avatars.prefix(4).enumerated()), id: \.offset) { index, initials in
                        FeedAvatar(
                            initials: initials,
                            tint: Self.attendeeTints[index % Self.attendeeTints.count],
                            size: 22
                        )
                        .overlay(
                            Circle().stroke(Theme.Color.appSurface, lineWidth: 2)
                        )
                    }
                }
                Text("+ \(strip.goingCount) going")
                    .font(.system(size: 11, weight: .medium))
                    .foregroundStyle(Theme.Color.appTextSecondary)
                Spacer(minLength: Spacing.s1)
                if let onRSVP {
                    Button(action: onRSVP) {
                        HStack(spacing: Spacing.s1) {
                            Icon(
                                strip.userIsGoing ? .check : .plusCircle,
                                size: 10,
                                strokeWidth: 3,
                                color: Theme.Color.magic
                            )
                            Text(strip.userIsGoing ? "Going" : "RSVP")
                                .font(.system(size: 11, weight: .bold))
                                .foregroundStyle(Theme.Color.magic)
                        }
                        .padding(.horizontal, Spacing.s3)
                        .frame(height: 26)
                        .background(Theme.Color.magicBg)
                        .clipShape(Capsule())
                    }
                    .buttonStyle(.plain)
                    .accessibilityLabel(strip.userIsGoing ? "Going" : "RSVP")
                    .accessibilityIdentifier("pulseRSVP_\(content.id)")
                }
            }
            .padding(.top, Spacing.s2)
        }
    }

    private var reactionStrip: some View {
        HStack(alignment: .center, spacing: 14) {
            ForEach(content.reactions) { reaction in
                reactionPill(reaction)
            }
            Spacer()
            HStack(spacing: Spacing.s1) {
                Icon(.messageCircle, size: 12, color: Theme.Color.appTextSecondary)
                Text(content.commentCount > 0 ? "Reply \(content.commentCount)" : "Reply")
                    .font(.system(size: 11.5, weight: .medium))
                    .foregroundStyle(Theme.Color.appTextSecondary)
            }
            .accessibilityElement()
            .accessibilityLabel(
                content.commentCount > 0
                    ? "Reply. \(content.commentCount) comments"
                    : "Reply"
            )
        }
        .padding(.top, Spacing.s2)
    }

    /// Cycling tint palette for the Event attendee mini-avatars (decorative;
    /// the live feed has no per-attendee identity, fixtures echo the design).
    private static let attendeeTints: [FeedAvatarTint] = [.orange, .sky, .violet, .green]

    @ViewBuilder
    private func reactionPill(_ reaction: PulseReaction) -> some View {
        let active = reaction.kind == content.reactions.first?.kind && content.userHasReacted
        if reaction.isInteractive {
            Button(action: onPrimaryReaction) { reactionLabel(reaction, active: active) }
                .buttonStyle(.plain)
                .accessibilityLabel("\(reaction.label.isEmpty ? "React" : reaction.label), \(reaction.count)")
                .accessibilityIdentifier("pulseReaction_\(content.id)_\(reaction.id.rawValue)")
        } else {
            reactionLabel(reaction, active: false)
                .accessibilityElement()
                .accessibilityLabel("\(reaction.label.isEmpty ? "Count" : reaction.label), \(reaction.count)")
        }
    }

    private func reactionLabel(_ reaction: PulseReaction, active: Bool) -> some View {
        HStack(spacing: Spacing.s1) {
            Icon(reaction.icon, size: 12, color: active ? Theme.Color.primary600 : Theme.Color.appTextSecondary)
            HStack(spacing: 3) {
                if !reaction.label.isEmpty {
                    Text(reaction.label)
                        .font(.system(size: 11.5, weight: .medium))
                        .foregroundStyle(active ? Theme.Color.primary600 : Theme.Color.appTextSecondary)
                }
                Text("\(reaction.count)")
                    .font(.system(size: 11.5))
                    .foregroundStyle(active ? Theme.Color.primary600 : Theme.Color.appTextSecondary)
            }
        }
    }

    private var a11yLabel: String {
        var parts = [content.authorName, content.intent.cardChipLabel]
        if let title = content.title, !title.isEmpty { parts.append(title) }
        if !content.body.isEmpty { parts.append(content.body) }
        if !content.media.isEmpty {
            parts.append("\(content.media.count) attached \(content.media.count == 1 ? "photo" : "photos")")
        }
        return parts.joined(separator: ". ")
    }
}

/// Right-aligned colored chip in the post header. Resolves intent →
/// foreground/background tokens against the existing design system.
public struct PulseIntentChip: View {
    private let intent: PulseIntent

    public init(intent: PulseIntent) {
        self.intent = intent
    }

    public var body: some View {
        HStack(spacing: Spacing.s1) {
            Icon(intent.icon, size: 10, strokeWidth: 2.5, color: foreground)
            Text(intent.cardChipLabel.uppercased())
                .font(.system(size: 10, weight: .bold))
                .tracking(0.4)
                .foregroundStyle(foreground)
        }
        .padding(.horizontal, Spacing.s2)
        .padding(.vertical, 2)
        .background(background)
        .clipShape(Capsule())
        .accessibilityLabel("\(intent.label) post")
    }

    private var foreground: Color {
        switch intent {
        case .all: Theme.Color.appTextSecondary
        case .ask: Theme.Color.warmAmber
        case .recommend: Theme.Color.success
        case .event: Theme.Color.magic
        case .lost: Theme.Color.rose
        case .announce: Theme.Color.slate
        }
    }

    private var background: Color {
        switch intent {
        case .all: Theme.Color.appSurfaceSunken
        case .ask: Theme.Color.warmAmberBg
        case .recommend: Theme.Color.successLight
        case .event: Theme.Color.magicBg
        case .lost: Theme.Color.roseBg
        case .announce: Theme.Color.slateBg
        }
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/feed/pulse/PulsePostCard.kt
@file:Suppress("MagicNumber", "PackageNaming", "LongMethod")

package app.pantopus.android.ui.screens.feed.pulse

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.pantopus.android.ui.components.AvatarWithIdentityRing
import app.pantopus.android.ui.components.IdentityPillar
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusIcon
import app.pantopus.android.ui.theme.PantopusIconImage
import app.pantopus.android.ui.theme.Radii
import app.pantopus.android.ui.theme.Spacing

/** VM-prepared content for a single Pulse card. */
@Immutable
data class PulsePostCardContent(
    val id: String,
    val authorName: String,
    val authorInitials: String,
    val authorVerified: Boolean,
    val meta: String,
    val intent: PulseIntent,
    val title: String?,
    val body: String,
    val reactions: List<PulseReaction>,
    val attendees: PulseAttendeeStrip?,
    val userHasReacted: Boolean,
)

/** Event card attendee strip — stacked avatars + going count + RSVP CTA. */
@Immutable
data class PulseAttendeeStrip(
    val avatars: List<String>,
    val goingCount: Int,
    val userIsGoing: Boolean,
)

/**
 * Pulse post card — entirely render-only; tap dispatch is parent-controlled.
 */
@Composable
fun PulsePostCard(
    content: PulsePostCardContent,
    onTap: () -> Unit,
    onPrimaryReaction: () -> Unit,
    onRSVP: (() -> Unit)? = null,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier =
            modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(Radii.lg))
                .border(1.dp, PantopusColors.appBorder, RoundedCornerShape(Radii.lg))
                .background(PantopusColors.appSurface)
                .clickable(onClick = onTap)
                .padding(Spacing.s3)
                .semantics(mergeDescendants = true) {
                    contentDescription = buildA11yLabel(content)
                }
                .testTag("pulsePostCard_${content.id}"),
        verticalArrangement = Arrangement.spacedBy(Spacing.s2),
    ) {
        CardHeader(content = content)
        if (!content.title.isNullOrEmpty()) {
            Text(
                text = content.title,
                fontSize = 13.5.sp,
                fontWeight = FontWeight.SemiBold,
                color = PantopusColors.appText,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
        }
        if (content.body.isNotEmpty()) {
            Text(
                text = content.body,
                fontSize = 12.5.sp,
                color = PantopusColors.appTextStrong,
                maxLines = if (!content.title.isNullOrEmpty()) 2 else 3,
                overflow = TextOverflow.Ellipsis,
            )
        }
        content.attendees?.let { attendees ->
            AttendeeStrip(attendees = attendees, onRSVP = onRSVP, postId = content.id)
        }
        ReactionStrip(
            content = content,
            onPrimary = onPrimaryReaction,
        )
    }
}

@Composable
private fun CardHeader(content: PulsePostCardContent) {
    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(9.dp),
    ) {
        AvatarWithIdentityRing(
            name = content.authorInitials,
            identity = IdentityPillar.Personal,
            ringProgress = if (content.authorVerified) 1f else 0.35f,
            size = 32.dp,
        )
        Column(modifier = Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(2.dp)) {
            Text(
                text = content.authorName,
                fontSize = 13.sp,
                fontWeight = FontWeight.SemiBold,
                color = PantopusColors.appText,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
            Text(
                text = content.meta,
                fontSize = 10.5.sp,
                color = PantopusColors.appTextSecondary,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
        }
        PulseIntentChip(intent = content.intent)
    }
}

@Composable
private fun AttendeeStrip(
    attendees: PulseAttendeeStrip,
    onRSVP: (() -> Unit)?,
    postId: String,
) {
    Column {
        Box(
            modifier =
                Modifier
                    .fillMaxWidth()
                    .height(1.dp)
                    .background(PantopusColors.appBorderSubtle),
        )
        Row(
            modifier = Modifier.fillMaxWidth().padding(top = Spacing.s2),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(Spacing.s2),
        ) {
            // Stacked avatars
            Box {
                attendees.avatars.take(4).forEachIndexed { index, initials ->
                    Box(
                        modifier =
                            Modifier
                                .padding(start = (index * 14).dp)
                                .size(22.dp)
                                .clip(CircleShape)
                                .border(2.dp, PantopusColors.appSurface, CircleShape),
                    ) {
                        AvatarWithIdentityRing(
                            name = initials,
                            identity = IdentityPillar.Personal,
                            ringProgress = 1f,
                            size = 22.dp,
                        )
                    }
                }
            }
            Text(
                text = "+ ${attendees.goingCount} going",
                fontSize = 11.sp,
                fontWeight = FontWeight.Medium,
                color = PantopusColors.appTextSecondary,
            )
            Spacer(modifier = Modifier.weight(1f))
            if (onRSVP != null) {
                Row(
                    modifier =
                        Modifier
                            .height(26.dp)
                            .clip(RoundedCornerShape(Radii.pill))
                            .background(PantopusColors.businessBg)
                            .clickable(onClick = onRSVP)
                            .padding(horizontal = 12.dp)
                            .testTag("pulseRSVP_$postId"),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(4.dp),
                ) {
                    PantopusIconImage(
                        icon = if (attendees.userIsGoing) PantopusIcon.Check else PantopusIcon.PlusCircle,
                        contentDescription = null,
                        size = 10.dp,
                        tint = PantopusColors.business,
                    )
                    Text(
                        text = if (attendees.userIsGoing) "Going" else "RSVP",
                        fontSize = 11.sp,
                        fontWeight = FontWeight.Bold,
                        color = PantopusColors.business,
                    )
                }
            }
        }
    }
}

@Composable
private fun ReactionStrip(
    content: PulsePostCardContent,
    onPrimary: () -> Unit,
) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(14.dp),
    ) {
        content.reactions.forEach { reaction ->
            ReactionPill(
                reaction = reaction,
                active = content.userHasReacted && reaction.kind == content.reactions.firstOrNull()?.kind,
                onClick = if (reaction.isInteractive) onPrimary else null,
                postId = content.id,
            )
        }
        Spacer(modifier = Modifier.weight(1f))
        Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(4.dp)) {
            PantopusIconImage(
                icon = PantopusIcon.Send,
                contentDescription = null,
                size = 12.dp,
                tint = PantopusColors.appTextSecondary,
            )
            Text(
                text = "Reply",
                fontSize = 11.5.sp,
                fontWeight = FontWeight.Medium,
                color = PantopusColors.appTextSecondary,
            )
        }
    }
}

@Composable
private fun ReactionPill(
    reaction: PulseReaction,
    active: Boolean,
    onClick: (() -> Unit)?,
    postId: String,
) {
    val tint = if (active) PantopusColors.primary600 else PantopusColors.appTextSecondary
    val base =
        Modifier
            .semantics {
                contentDescription = "${reaction.label.ifEmpty { "Count" }}, ${reaction.count}"
            }
    val withClick =
        if (onClick != null) {
            base.clickable(onClick = onClick).testTag("pulseReaction_${postId}_${reaction.kind.key}")
        } else {
            base
        }
    Row(
        modifier = withClick,
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(4.dp),
    ) {
        PantopusIconImage(
            icon = reaction.icon,
            contentDescription = null,
            size = 12.dp,
            tint = tint,
        )
        if (reaction.label.isNotEmpty()) {
            Text(
                text = reaction.label,
                fontSize = 11.5.sp,
                fontWeight = FontWeight.Medium,
                color = tint,
            )
        }
        Text(
            text = "${reaction.count}",
            fontSize = 11.5.sp,
            color = tint,
        )
    }
}

/**
 * Right-aligned colored chip in the post header. Resolves intent →
 * foreground/background tokens against the existing design system.
 */
@Composable
fun PulseIntentChip(intent: PulseIntent) {
    val (fg, bg) = intent.tintColors()
    Row(
        modifier =
            Modifier
                .clip(RoundedCornerShape(Radii.pill))
                .background(bg)
                .padding(horizontal = 8.dp, vertical = 2.dp)
                .semantics { contentDescription = "${intent.label} post" },
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(4.dp),
    ) {
        PantopusIconImage(
            icon = intent.icon,
            contentDescription = null,
            size = 10.dp,
            tint = fg,
        )
        Text(
            text = intent.cardChipLabel.uppercase(),
            fontSize = 10.sp,
            fontWeight = FontWeight.Bold,
            color = fg,
        )
    }
}

private fun PulseIntent.tintColors(): Pair<Color, Color> =
    when (this) {
        PulseIntent.All -> PantopusColors.appTextSecondary to PantopusColors.appSurfaceSunken
        PulseIntent.Ask -> PantopusColors.warning to PantopusColors.warningBg
        PulseIntent.Recommend -> PantopusColors.success to PantopusColors.successBg
        PulseIntent.Event -> PantopusColors.business to PantopusColors.businessBg
        PulseIntent.Lost -> PantopusColors.error to PantopusColors.errorBg
        PulseIntent.Announce -> PantopusColors.appTextStrong to PantopusColors.appSurfaceSunken
    }

private fun buildA11yLabel(content: PulsePostCardContent): String {
    val parts = mutableListOf<String>()
    parts.add(content.authorName)
    if (content.intent.cardChipLabel.isNotEmpty()) parts.add(content.intent.cardChipLabel)
    content.title?.takeIf { it.isNotEmpty() }?.let { parts.add(it) }
    if (content.body.isNotEmpty()) parts.add(content.body)
    return parts.joinToString(". ")
}
